docs/adr/adr-013-bootstrap-from-previous-peers.md

```diff
@@ -0,0 +1,370 @@
+# ADR 013: Bootstrap from previous peers
+
+## Changelog
+
+* 2023-05-04: initial draft
+* 2023-05-05: update peer selection with new simpler solution + fix language
+
+## Context
+
+Celestia node relies on a set of peer to peer protocols that assume a decentralized network, where nodes are connected to each other and can exchange data.
+
+Currently, when a node joins/comes up on the network, it relies, by default, on hardcoded bootstrappers for 3 things:
+
+1. initializing its store with the first (genesis) block
+2. requesting the network head to which it sets its initial sync target, and
+3. bootstrapping itself with some other peers on the network which it can utilise for its operations.
+
+This is centralisation bottleneck as it happens both during initial start-up and for any re-starts.
+
+In this ADR, we wish to aleviate this problem by allowing nodes to bootstrap from previously seen peers. This is a simple solution that allows the network alleviate the start-up centralisation bottleneck such that nodes can join the network without the need for hardcoded bootstrappers up re-start
+
+(_Original issue available in references_)
+
+## Decision
+
+Periodically store the addresses of previously seen peers in a key-value database, and allow the node to bootstrap from these peers on start-up.
+
+## Detailed Design
+
+### What systems will be affected?
+
+* `peerTracker` and `libhead.Exchange` from `go-header`
+* `newInitStore` in `nodebuilder/header.go`
+
+### Database
+
+We will use a badgerDB datastore to store the addresses of previously seen peers. The database will be stored in the `data` directory, and will be named `peers.db`. The database will have a single key-value pair, where the key is the peer ID, and the value is the peer multiaddress.
+
+The peer store will implement the following interface:
+
+```go
+type PeerStore interface {
+   // Put adds a list of peers to the store.
+    Put([]peer.AddrInfo) error
+    // Get returns a peer from the store.
+    Load() ([]peer.AddrInfo, error)
+    // Delete removes a peer from the store.
+    Delete(peer.ID) error
+}
+```
+
+And we expect the underlying implementation to use a badgerDB datastore. Example:
+
+```go
+type peerStore struct {
+    db datastore.Datastore
+}
+```
+
+### Peer Selection & Persistence
+
+To periodically select the _"good peers"_ that we're connected to and store them in the peer store, we will rely on the internals of `go-header`, specifically the `peerTracker` and `libhead.Exchange` structs.
+
+`peerTracker` has internal logic that continuiously tracks and garbage collects peers, so that if new peers connect to the node, peer tracker keeps track of them in an internal list, as well as it marks disconnected peers for removal. `peerTracker` also blocks peers that fail to answer*.
```

Comment:
```suggestion
`peerTracker` has internal logic that continuiously tracks and garbage collects peers, so that if new peers connect to the node, peer tracker keeps track of them in an internal list, as well as it marks disconnected peers for removal. `peerTracker` also blocks peers that behave maliciously*.
```